File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        
        Queue<Word> pq = new LinkedList<>();
        pq.add(new Word(begin, 0));
        boolean[] visited = new boolean[words.length];
        
        while(!pq.isEmpty()) {
            Word curr = pq.poll();
            
            if (curr.word.equals(target)) {
                answer = curr.count;
                break;
            }
            
            for (int i = 0; i < words.length; i++) {
                
                if (visited[i] || getDifferentWordNum(curr.word, words[i]) != 1) {
                    continue;
                } 
                
                visited[i] = true;
                pq.add(new Word(words[i], curr.count+1));
            }
        }
        
        return answer;
    }
    
    /* 다른 문자의 개수 반환 */
    private int getDifferentWordNum(String first, String second) {
        int count = 0;
        for (int i = 0; i < first.length(); i++) {
            if (first.charAt(i) != second.charAt(i)) {
                count++;
            }
        }
        return count;
    }
}

class Word {
    String word;
    int count;
    public Word(String word, int count) {
        this.word = word;
        this.count = count;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

//8:48
struct Word {
public:
    string word;
    int step;
    
    Word(string word, int step) {
        this->word = word;
        this->step = step;
    }
};

int solution(string begin, string target, vector<string> words) {
    
    queue<Word> q; // 문자 큐
    q.push(Word(begin, 0));
    
    vector<bool> visited(words.size());
    
    while (!q.empty()) {
        
        Word curr = q.front();
        q.pop();
        
        if (curr.word == target) {
            return curr.step;
        }
        
        for (int i = 0; i < words.size(); i++) {
            
            if (visited[i]) {
                continue;
            }
            
            int count = begin.size();
            for (int j = 0; j < begin.size(); j++) {
        
                if (words[i][j] == curr.word[j]) {
                    count--;
                }
            }
        
            if (count == 1) {
                visited[i] = true;
                q.push(Word(words[i], curr.step+1));
            }
        }
    }
    
    return 0;
}